File: sam-app/japicc-check/japicc/japicc.go
package japicc

import (
	"bytes"
	"errors"
	"log"
	"os"
	"os/exec"
)

// path to JAPICC command on local system
const binary = "/usr/bin/japi-compliance-checker"

// IsInstalled Check if JAPICC is installed.
func IsInstalled() bool {
	_, err := os.Stat(binary)
	if err == nil {
		return true
	}
	return false
}

// GetVersion Run JAPICC with option '--version' and return the output.
func GetVersion() ([]byte, error) {
	return runCommand("--version")
}

func Check(oldFilePath string, newFilePath string, reportFilePath string) error {

	_, err := runCommand("-old", oldFilePath, "-new", newFilePath, "-report-path", reportFilePath)
	if err != nil {
		var exitError *exec.ExitError
		if errors.As(err, &exitError) {
			if exitError.ExitCode() == 1 {
				// incompatibilities found
			} else {
				return err
			}
		} else {
			return err
		}
	}

	return nil
}

// runCommand Run JAPICC with given arguments.
func runCommand(args ...string) ([]byte, error) {

	// check if JAPICC is installed
	if !IsInstalled() {
		return nil, errors.New("JAPICC is not installed.")
	}

	// prepare JAPICC command
	cmd := exec.Command(binary, args...)

	// buffer output sent to STDOUT and STDERR
	var out bytes.Buffer
	cmd.Stdout = &out
	cmd.Stderr = &out

	// execute JAPICC command
	log.Println("Run JAPICC command:", args)
	err := cmd.Run()

	// log command output
	data := out.Bytes()
	log.Println("Output:\n", string(data))

	return data, err
}

File: sam-api/japicc-check/src/main/java/org/jarhc/online/japicc/JAPICC.java
package org.jarhc.online.japicc;

import com.amazonaws.xray.AWSXRay;
import com.amazonaws.xray.entities.Subsegment;
import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.InputStream;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

class JAPICC {

	private static final Logger logger = LoggerFactory.getLogger(JAPICC.class);

	private static final String JAPI_COMPLIANCE_CHECKER = "/usr/bin/japi-compliance-checker";

	void execute(File oldVersionFile, File newVersionFile, File reportFile) {
		try (Subsegment xray = AWSXRay.beginSubsegment("JAPICC.execute")) {

			int exitCode;
			try {

				// run JAPICC
				Process process = new ProcessBuilder()
						.command(
								JAPI_COMPLIANCE_CHECKER,
								"-old", oldVersionFile.getAbsolutePath(),
								"-new", newVersionFile.getAbsolutePath(),
								"-report-path", reportFile.getAbsolutePath()
						)
						.start();

				// wait for process to finish
				exitCode = process.waitFor();

				// read output from STDOUT
				try (InputStream inputStream = process.getInputStream()) {
					ByteArrayOutputStream buffer = new ByteArrayOutputStream();
					inputStream.transferTo(buffer);
					String output = buffer.toString();
					logger.info("JAPICC output:\n{}", output);
				}

				// read output from STDERR
				try (InputStream errorStream = process.getErrorStream()) {
					ByteArrayOutputStream buffer = new ByteArrayOutputStream();
					errorStream.transferTo(buffer);
					String output = buffer.toString();
					if (!output.isEmpty()) {
						logger.error("JAPICC error:\n{}", output);
					}
				}

			} catch (Exception e) {
				xray.addException(e);
				throw new RuntimeException("Error running JAPICC.", e);
			}

			// handle exit code
			switch (exitCode) {
				case 0: // OK (no incompatibilities)
				case 1: // OK (incompatibilities found)
					break;
				default:
					throw new RuntimeException("JAPICC failed with exit code " + exitCode + ".");
			}

		}
	}
}
